### PR TITLE
Add min / max values to virtual column metadata to allow segment pruning

### PR DESCRIPTION
**Pull request by saurabhd336**
When running queries with predicates on $segmentName, segments are not pruned due to missing min / max value metadata. This PR adds that to all virtual columns. Ideally this is something that can be handled as a broker side pruner too (will raise a followup PR for that).

### CODE REVIEW

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/DefaultNullValueVirtualColumnProvider.java [hunk 1]
```diff
@@ -88,6 +88,11 @@ public ColumnMetadataImpl buildMetadata(VirtualColumnContext context) {
     ColumnMetadataImpl.Builder builder = getColumnMetadataBuilder(context).setCardinality(1).setHasDictionary(true);
     if (context.getFieldSpec().isSingleValueField()) {
       builder.setSorted(true);
+
+      if (context.getFieldSpec().getDefaultNullValue() instanceof Comparable) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/DefaultNullValueVirtualColumnProvider.java:92]: (minor) Put `context.getFieldSpec().getDefaultNullValue()` into a local variable to be reused

Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/DefaultNullValueVirtualColumnProvider.java:92]: For BYTES column, we want to wrap the value into a `ByteArray` object and set it as the min & max value

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/DefaultNullValueVirtualColumnProvider.java [hunk 2]
```diff
@@ -88,6 +88,11 @@ public ColumnMetadataImpl buildMetadata(VirtualColumnContext context) {
...
+        builder.setMinValue((Comparable) context.getFieldSpec().getDefaultNullValue());
+        builder.setMaxValue((Comparable) context.getFieldSpec().getDefaultNullValue());
+      }
     } else {
       // When there is no value for a multi-value column, the maxNumberOfMultiValues and cardinality should be
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/DefaultNullValueVirtualColumnProvider.java:97]: We want to set the same min & max value for the MV column

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/DefaultNullValueVirtualColumnProvider.java (hunk 1 above)
Comment (saurabhd336) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/DefaultNullValueVirtualColumnProvider.java:92]: Ack

Comment (saurabhd336) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/DefaultNullValueVirtualColumnProvider.java]: Ack

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/DefaultNullValueVirtualColumnProvider.java (hunk 2 above)
Comment (saurabhd336) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/DefaultNullValueVirtualColumnProvider.java]: Ack. Added for all column types

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/DefaultNullValueVirtualColumnProvider.java [hunk 3]
```diff
@@ -94,6 +95,36 @@ public ColumnMetadataImpl buildMetadata(VirtualColumnContext context) {
...
+        break;
+      case STRING:
+        builder.setMinValue((String) defaultNullValue).setMaxValue((String) defaultNullValue);
+        break;
+      case BYTES:
+        builder.setMinValue(new ByteArray((byte[]) defaultNullValue))
+            .setMaxValue(new ByteArray((byte[]) defaultNullValue));
+        break;
+      default:
+        break;
```
Comment (GSharayu) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/DefaultNullValueVirtualColumnProvider.java:125]: (minor nit) redundant break

**Review by Jackie-Jiang — APPROVED** [review]

**Pull request merged**